contrib/smtp_proxy/filters/network/source/smtp_decoder.h

```diff
@@ -0,0 +1,87 @@
+#pragma once
+#include <cstdint>
+
+#include "envoy/common/platform.h"
+
+#include "source/common/buffer/buffer_impl.h"
+#include "source/common/common/logger.h"
+
+#include "absl/container/flat_hash_map.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace SmtpProxy {
+
+// Smtp message decoder.
+class Decoder {
+public:
+  virtual ~Decoder() = default;
+
+  // The following values are returned by the decoder, when filter
+  // passes bytes of data via onData method:
+  enum class Result {
+    ReadyForNext, // Decoder processed previous message and is ready for the next message.
+    NeedMoreData, // Decoder needs more data to reconstruct the message.
+    Bad
```

Comment:
Maybe `Bad` should be changed to something more meaningful like `DecodingError`. WDYT?

Reply:
renamed ProtocolError